rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,338 @@
+  teleportConfig:
+    auth_service:
+      authentication:
+        connector_name: "my-connector"
+proxy:
+  teleportConfig:
+    proxy_service:
+      acme:
+        enabled: true
+        email: foo@example.com
+```
+
+The following values are core values: users must set them for the chart to work
+properly. They support the happy path. Those values should not be changed by
+the proposal as it would harm backward compatibility and user experience.
+
+- `clusterName`
+- `publicAddr`
+- `chartMode`
+- `aws`
+- `gcp`
+- `enterprise`
+- `operator`
+
+The following values are used to generate Teleport's configuration, we must
+continue to support them for backward compatibility, but using
+`*.teleportConfig` should be preferred.
+
+- `kubeClusterName`
+- `authentication`
+- `authenticationType`
+- `authenticationSecondFactor`
+- `proxyListenerMode`
+- `sessionRecording`
+- `separatePostgresListener`
+- `separateMongoListener`
+- `kubePublicAddr`
+- `mongoPublicAddr`
+- `mysqlPublicAddr`
+- `postgresPublicAddr`
+- `sshPublicAddr`
+- `tunnelPublicAddr`
+- `acme`
+- `acmeEmail`
+- `acmeURI`
+- `log`
+
+#### Chart-specific configuration values
+
+Some values are used to configure the Kubernetes resources deploying Teleport.
+When specified they should apply to both auth and proxy deployments.  Those
+values are:
+
+- `podSecurityPolicy`
+- `labels`
+- `highAvailability`
+- `tls`
+- `image`
+- `enterpriseImage`
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `imagePullPolicy`
+- `initContainers`
+- `postStart`
+- `securityContext`
+- `priorityClassName`
+- `tolerations`
+- `probeTimeoutSeconds`
+- `teleportVersionOverride`
+- `resources`
+
+A few values will have to be treated differently:
+
+- `persistence` will only apply to the `auth` deployment
+- `service` will only apply to the `proxy` service
+- `serviceAccount.name` will apply to the `auth`, the proxy service account
+  name should be the auth one suffixed with `-proxy`
+
+Some users will need to set different values for auth and proxy pods, the
+following values should be also available under `auth` and `proxy`. Those
+specific values should take precedence over the ones at the root.
+
+- `labels`
+- `highAvailability` (except the certManager section)
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `initContainers`
+- `postStart`
+- `tolerations`
+- `teleportVersionOverride`
+- `resources`
+
+### Update strategy between major versions
+
+Auth pods have to be updated before proxies.  Helm does not support applying
+resources in a specific order.
+
+Both auth and proxy rollouts will be triggered at the same time, but the proxy
+one should be held until all auth pods are rolled out. Not waiting for the full
+rollout will cause the load to spread unevenly across auth pods, which will be
+harmful at scale.
+
+Proxies will have an initContainer checking if all auth pods from the past
```

Comment:
I would assume that most of them would run with `--atomic`
```--atomic                                     if set, upgrade process rolls back changes made in case of failed upgrade. The --wait flag will be set automatically if --atomic is used```

Reply:
In this case yes, we must warn the users the rollout can take some time and they should adjust their timeouts. Releasing will be longer than before if we setup disruption mitigations like preStop hooks, graceful shutdowns and sequentially update auth and proxies.

Reply:
Added a section about documenting the possible rollout time and timeout tuning in the RFD: **https://github.com/gravitational/teleport/pull/18274/commits/ec8823cc82ab5a41f8861a8be0e344eab74aea85**